src/circuit/circuit_frame.rs

```diff
@@ -2246,94 +2191,283 @@ fn apply_continuation<F: PrimeField, CS: ConstraintSystem<F>>(
             )?;
 
             let continuation_is_tail = alloc_equal(
-                &mut cs.namespace(|| "call2 continuation is tail using newer continuation"),
+                &mut cs.namespace(|| "Call2: continuation is tail"),
                 continuation.tag(),
                 &g.tail_cont_tag,
             )?;
 
             let tail_cont = AllocatedContPtr::pick(
-                &mut cs.namespace(|| "call2 the tail continuation using newer continuation"),
+                &mut cs.namespace(|| "Call2: the tail continuation"),
                 &continuation_is_tail,
                 &continuation,
                 &newer_cont,
             );
 
-            (body_form, newer_env, tail_cont)
+            let result_is_fun = alloc_equal(
+                cs.namespace(|| "Call2: result_is_fun"),
+                function.tag(),
```

Comment:
The naming implies we are checking `result`, but we're actually looking at `function`. As far as I can tell, this is bound when destructuring a `Call` continuation. However, it's not clear that `Call` and `Call0` hash components are laid out the same (I think they're not).

I'm not saying this is wrong (I can't tell), but it might be. If not, let's try to clarify naming/structure so it's clear what's happening and why.

If it already is and I'm just misreading, sorry — maybe the explanation that helps me understand will at least be a useful comment for the future, though.

Reply:
Good catch! This was wrong indeed.